schnorr_fun/src/share_backup.rs

```diff
@@ -0,0 +1,238 @@
+            }
+            FrostBackupDecodeError::ShareIndexIsZero => {
+                write!(
+                    f,
+                    "Can not have a share index of zero since this immediately implies the secret.",
+                )
+            }
+        }
+    }
+}
+
+/// Create an identifier that's used to determine compatibility of shamir secret shares.
+/// The first 4 bech32 chars from a hash of the polynomial coefficients.
+/// Collision expected once in (32)^4 = 2^20.
+pub fn polynomial_identifier<H: Default + Digest<OutputSize = U32>>(
+    polynomial: Vec<Point>,
+) -> [u5; 4] {
+    let hash = H::default();
+    hash.add(&polynomial[..]).finalize().to_base32()[0..4]
+        .try_into()
+        .expect("4 bech32 chars must fit 4 character arry")
+}
+
+/// Create a bech32m secret share backup
+///
+/// Requires that the threshold be no greater than 1024.
+/// If using an integer (small) index, the backup will be 67 bech32 characters.
+/// If using a scalar index, the backup will be 118 bech32 characters.
+pub fn encode_backup<H: Default + Digest<OutputSize = U32>>(
+    polynomial: &Vec<Point>,
+    secret_share: &Scalar,
+    share_index: &Scalar<Public>,
+) -> String {
+    if poly::point::eval(polynomial, *share_index) != g!(secret_share * G) {
+        panic!("Secret share is not valid with respect to the polynomial!")
+    }
+
+    let threshold = polynomial.len();
+    let mut data = [u5::default(); 2 + 4 + 52 + 52];
+
+    if threshold > 1024 {
+        panic!("Polynomial has too high of a threshold, {threshold} > 1024");
+    }
+    let threshold_u5 = (threshold - 1).to_le_bytes().to_vec().to_base32();
+    for (i, byte) in threshold_u5.into_iter().take(2).enumerate() {
+        data[i] = byte;
+    }
+
+    let polynomial_identifier_u5 = polynomial_identifier::<H>(polynomial.clone());
+    for (i, byte) in polynomial_identifier_u5.into_iter().enumerate() {
+        data[2 + i] = byte;
+    }
+
+    let secret_share_u5 = secret_share.to_bytes().to_vec().to_base32();
+    for (i, byte) in secret_share_u5.into_iter().enumerate() {
+        data[2 + 4 + i] = byte;
+    }
+
+    let is_small =
+        share_index.to_bytes()[0..31].iter().all(|b| *b == 0) && share_index.to_bytes()[31] < 32;
+
+    let n_unused_bytes = if is_small {
+        let share_index_u5 =
+            u5::try_from_u8(share_index.to_bytes()[31]).expect("must be less than 32");
+        data[2 + 4 + 52] = share_index_u5;
+        52 - 1
+    } else {
+        let share_index_u5 = share_index.to_bytes().to_base32();
+        for (i, byte) in share_index_u5.iter().enumerate() {
+            data[2 + 4 + 52 + i] = *byte;
+        }
+        0
+    };
+
+    bech32::encode("frost", &data[..(data.len() - n_unused_bytes)], Bech32m)
+        .expect("hrp must be valid")
+}
+
+/// Decode a bech32m secret share backup
+pub fn decode_backup(
+    encoded: String,
+) -> Result<(usize, [u5; 4], Scalar, Scalar), FrostBackupDecodeError> {
```

Comment:
Make a struct for the return value here so it's easier to tell what each field means. My plan would be:

1. Make a struct `ShareBackup` implement display on it which should product the actual string.
2. When you decode the string you get `ShareBackup`
3. You have  a `ShareBackup::new` which takes the same arguments as the encode function now.
4. You can also have a floating encode function for convenience too.

I know you originally had something more like this and I suggested single functions. I didn't realize at that time the decode return value would be so complicated. Sorry!